lib/cog/command/service/data_store.ex

```diff
@@ -0,0 +1,81 @@
+defmodule Cog.Command.Service.DataStore do
+  @moduledoc """
+  Stores an arbitrary data structure for a given key. The only requirement is
+  that the data structure must be able to be encoded as JSON.
+
+  Keys may be fetched, replaced, and deleted.
+
+  The JSON data is stored on the filesystem of the Cog host. See the
+  Cog.NestedFile module for more details.
+  """
+
+  use GenServer
+
+  alias Cog.NestedFile
+
+  defstruct [:base_path]
+
+  @doc """
+  Starts the #{inspect __MODULE__} service. Accepts a path to use for the
+  base directory to store content under.
+  """
+  def start_link(base_path),
+    do: GenServer.start_link(__MODULE__, base_path, name: __MODULE__)
+
+  @doc """
+  Fetches the given key. Returns `{:ok, value}` if the key exists or `{:error,
+  :unknown_key}` if it doesn't exist.
+  """
+  def fetch(namespace, key),
+    do: GenServer.call(__MODULE__, {:fetch, namespace, key})
+
+  @doc """
+  Replaces or sets the given key with the value. Returns `{:ok, value}`.
+  """
+  def replace(namespace, key, value),
+    do: GenServer.call(__MODULE__, {:replace, namespace, key, value})
+
+  @doc """
+  Deletes the given key. Returns `{:ok, key}` when successfully
+  deleted or `{:error, :unknown_key}` if it doesn't exist.
+  """
+  def delete(namespace, key),
+    do: GenServer.call(__MODULE__, {:delete, namespace, key})
+
+  def init(nil),
+    do: raise "Unable to start #{__MODULE__}: Data path not configured"
+  def init(base_path) do
+    state = %__MODULE__{base_path: base_path}
```

Comment:
The `NestedFile` module will try to create it if it does not exist the first time it is attempted to be written to and return an error if it is unable to. That felt like the right thing in this case since not everyone is likely to use the commands that make use of it and I wouldn't want to keep Cog from working. It also defaults to be nested under the default data directory just like logs and audit_logs so it feels pretty safe to me.

Happy to hear arguments to the contrary if you think we should do something else.